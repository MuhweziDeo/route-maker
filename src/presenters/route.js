import React, { useRef, useEffect } from "react";
import {View, Text, StyleSheet, Dimensions, Button, ActivityIndicator} from "react-native";
import { styles as custom } from "./origin";
import MapView, { Marker, Polyline } from "react-native-maps";


export const RoutePresenter = ({ mode, setMode, destination, origin, coordinates, loading, error}) => {
    const ref = useRef(null);

    useEffect(() => {
        if(ref && ref.current) {
            setTimeout(() => {
                ref.current.fitToCoordinates([
                    { latitude: destination.geometry.lat, longitude: destination.geometry.lng },
                    { latitude: origin.geometry.lat, longitude: origin.geometry.lng }
                ], {edgePadding: { top: 70, right: 70, bottom: 70, left: 70}});
            },400)

        }
    }, [ref]);

    return (
        <View styles={custom.container}>
            <MapView
                ref={ref}
                style={styles.mapStyle}
                showsUserLocation={true}
             >
                <Marker
                    coordinate={{
                        latitude: destination.geometry.lat,
                        longitude: destination.geometry.lng
                    }}
                    pinColor="red"
                />

                { coordinates.length > 0 &&
                <Polyline
                    coordinates={[...coordinates]}
                    strokeColor="blue"
                    strokeWidth={3}
                />
                }
                <Marker
                    coordinate={{
                        latitude: origin.geometry.lat,
                        longitude: origin.geometry.lng
                    }}
                    pinColor="red"
                />
            </MapView>
          <View style={{padding:20, flexDirection: "row", justifyContent: "center"}}>
              <View style={[styles.margin]}>
                  <Button
                      disabled={loading}
                      onPress={() => setMode("car")}
                      color={mode === "car" ? "green": ""} title={"Driving"}/></View>
              <View style={[styles.margin]}>
                  <Button
                      disabled={loading}
                      color={mode === "bike" ? "green": ""}
                      onPress={() => setMode("bike")} title={"Bicycling"}/></View>
              <View style={[styles.margin]}>
                  <Button
                      disabled={loading}
                      color={mode === "truck" ? "green": ""}
                      onPress={() => setMode("truck")} title={"Transit"}/></View>
              <View style={[styles.margin]}>
                  <Button
                      disabled={loading}
                      color={mode === "foot" ? "green": ""}
                      onPress={() => setMode("foot")} title={"Walking"}/></View>
          </View>
            {loading && <ActivityIndicator size="large" color="#0000ff" />}
            { error &&
                <View style={{justifyContent: "space-around", alignItems: "center", marginBottom: 8}}>
                    <Text id={"error-message"} style={{fontWeight: "bold", color: 'red'}}>
                        Something Went wrong Please Try again Later
                    </Text>
                </View>
            }
            <View style={{justifyContent: "space-around", alignItems: "center", marginBottom: 8}}>
                <Text style={{fontWeight: "bold"}}>DESTINATION</Text>
                <Text>{destination.formatted}</Text>
            </View>

            <View style={{justifyContent: "space-around",alignItems: "center",marginBottom: 8 }}>
                <Text style={{fontWeight: "bold"}}>ORIGIN</Text>
                <Text>{origin.formatted}</Text>
            </View>

            <View style={{justifyContent: "space-around", alignItems: "center", marginBottom: 8 }}>
                <Text style={{fontWeight: "bold"}}>MODE</Text>
                <Text>{mode}</Text>
            </View>
        </View>
    )
};

const styles = StyleSheet.create({
    mapStyle: {
        width: Dimensions.get('window').width,
        height: Dimensions.get('window').height/1.7,
    },
    margin: {
        marginRight: 10
    }
});
