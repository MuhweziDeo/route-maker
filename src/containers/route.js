import React, { useState, useEffect, useContext } from "react";
import { RoutePresenter } from "../presenters";
import * as Location from "expo-location";
import { AppContext } from "../../App";
import { graphHopperApiKey, graphHopperUrl } from "../config";

export const RouteContainer = props => {
    const[location, setLocation] = useState({latitude: 0.3974545, longitude: 32.625096});
    const[coordinates, setCoordinates] = useState([]);
    const[mode, setMode] = useState("");
    const[loading, setLoading] = useState(false);
    const context = useContext(AppContext);
    const[error, setError] =useState(false);
    const loadLocation = async() => {
      const location = await Location.getCurrentPositionAsync();
      setLocation({latitude: location.coords.latitude,
          longitude: location.coords.longitude})
    };

    useEffect(() => {
        const load = async() => {
            await loadLocation();
        };
      return () =>  load();
    }, [location]);

    useEffect(() => {
        context.mode = mode;
        setLoading(true);
        fetch(`${graphHopperUrl}?point=${Number(context.destination.geometry.lat)},${Number(context.destination.geometry.lng)}&point=${context.origin.geometry.lat},${context.origin.geometry.lng}&vehicle=${mode}&key=${graphHopperApiKey}&type=json&points_encoded=false&elevation=true&turn_cost=false&weighting=fastest&locale=en-GB
        `).then((res) => res.json())
            .then(data => {
                const{paths} = data;
                const[{points: {coordinates}},] = paths;
                let points = coordinates.map(([longitude, latitude]) => {
                    return {
                        longitude,
                        latitude
                    }
                });
                setLoading(false);
                setCoordinates(points);
            })
            .catch(() => {
                setLoading(false);
                setError(true);
                setTimeout(() => {
                    setError(false)
                }, 5000)
            });
    }, [mode]);

    return (
        <RoutePresenter
            {...props}
            initialRegion={location}
            mode={mode}
            setMode={setMode}
            origin={context.origin}
            destination={context.destination}
            coordinates={coordinates}
            loading={loading}
            error={error}
        />
    )
};
