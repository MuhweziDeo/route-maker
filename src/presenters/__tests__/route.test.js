import 'react-native';
import React from "react";
import renderer from 'react-test-renderer'
import { RoutePresenter } from "../route";

export const RouterComponentProps = {
    navigation: {
        navigate: jest.fn()
    },

    initialRegion: {
        longitude: 0.0003,
        latitude: 32.000
    },
    mode: "transit",
    destination: {
        geometry: {
            lng: 0.44,
            lat: 32.00
        }
    },
    origin: {
        geometry: {
            lng: 0.44,
            lat: 32.00
        }
    }

};

jest.mock('react-native-maps', () => {
    const { View } = require('react-native');
    const MockMapView = (props) => {
        return <View>{props.children}</View>;
    };
    const MockMarker = (props) => {
        return <View>{props.children}</View>;
    };
    return {
        __esModule: true,
        default: MockMapView,
        Marker: MockMarker,
    };
});

describe('<OriginPresenter/>', () => {
    it("should match snapshot", () => {
        const wrapper = renderer.create(<RoutePresenter {...RouterComponentProps}/>);
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
});
