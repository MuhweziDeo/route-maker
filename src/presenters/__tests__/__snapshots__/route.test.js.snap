// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OriginPresenter/> should match snapshot 1`] = `
<View>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
            Object {
              "color": "red",
            },
          ]
        }
      >
        &lt; Destination
      </Text>
    </View>
  </View>
  <View>
    <View />
    <View />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "marginRight": 10,
          },
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Driving
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "marginRight": 10,
          },
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Bicycling
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "marginRight": 10,
          },
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "green",
                },
              ]
            }
          >
            Transit
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "marginRight": 10,
          },
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Walking
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-around",
      }
    }
  >
    <Text>
      Destination
    </Text>
    <Text />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-around",
      }
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Origin
    </Text>
    <Text />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-around",
      }
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Mode
    </Text>
    <Text>
      transit
    </Text>
  </View>
</View>
`;
